.github/workflows/ci.yml

```diff
@@ -23,6 +23,28 @@ env:
   MAKEJOBS: '-j10'
 
 jobs:
+  test-each-commit:
+    name: 'test each commit'
+    runs-on: ubuntu-22.04
+    if: github.event_name == 'pull_request' && github.event.pull_request.commits != 1
+    timeout-minutes: 360  # Use maximum time, see https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes. Assuming a worst case time of 1 hour per commit, this leads to a --max-count=6 below.
+    env:
+      MAX_COUNT: 6
+    steps:
+      - run: echo "FETCH_DEPTH=$((${{ github.event.pull_request.commits }} + 2))" >> "$GITHUB_ENV"
```

Comment:
I'm surprised this doesn't cause a failure on pull requests that build on other pull requests.

If pull request B has 3 commits that build on top of pull request A with 10 commits, then `FETCH_DEPTH` would be 5. I believe that should cause `git log --merges -1` below to fail, because it won't have the merge commit.

But looking at #28201 this seems to work fine. But when pull request A is on this repo and pull request B is on a fork repo it doesn't, it seems: https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2134851003

Reply:
@ryanofsky fixed a simpler case of this in #28573, where PR B has 1 commit. Though I'm confused why he ran into it while #28201 doesn't.

Reply:
> @ryanofsky fixed a simpler case of this in #28573, where PR B has 1 commit. Though I'm confused why he ran into it while #28201 doesn't.

I think there shouldn't be a problem here because even if a pull request is based on another pull request, the `github.event.pull_request.commits` number will include the number of commits in both pull requests, because github treats pull requests as being independent and isn't aware of relationships between them.

The case I was fixing in https://github.com/bitcoin/bitcoin/pull/28573 was a more strange case described in https://github.com/bitcoin/bitcoin/pull/28497#issuecomment-1743430631. Sometimes, when I have a PR `B` based on a PR `A`, I base the new commits in `B` on top of `A` merged with `master`, instead of just `A` like most other developers do. (I do this just because it lets me avoid needing to rebase A when B has a merge conflict.) This didn't work with the old test-each-commit implementation because the pushed branch had a extra merge commit that the script was not expecting, and broke it.

Reply:
> because github treats pull requests as being independent and isn't aware of relationships between them.

On the fork I tested on the pull request wasn't against `master` but against branch `A`.